backend/module/chaos/serverexperimentation/xds/ecds.go

```diff
@@ -0,0 +1,255 @@
+	var downstreamCluster string
+	var upstreamCluster string
+	var isExternalFault = false
+
+	abort := DefaultAbortFaultConfig
+	delay := DefaultDelayFaultConfig
+
+	for _, experiment := range experiments {
+		httpFaultConfig := &serverexperimentation.HTTPFaultConfig{}
+		if !maybeUnmarshalFaultTest(experiment, httpFaultConfig) {
+			continue
+		}
+
+		var err error
+		upstreamCluster, downstreamCluster, err = getClusterPair(httpFaultConfig)
+		if err != nil {
+			logger.Errorw("Invalid http fault config", "config", httpFaultConfig)
+			continue
+		}
+
+		isExternalFault, abort, delay, err = createAbortDelayConfig(httpFaultConfig, downstreamCluster)
+		if err != nil {
+			logger.Errorw("Unable to create ECDS filter fault", "config", httpFaultConfig)
+			continue
+		}
+
+		// We can only perform one fault test per cluster with ECDS. Hence we break.
+		break
+	}
+
+	faultFilter := &gcpFilterFault.HTTPFault{
+		Delay: delay,
+		Abort: abort,
+
+		// override runtimes so that default runtime is not used.
+		DelayPercentRuntime:    delayPercentRuntime,
+		DelayDurationRuntime:   delayDurationRuntime,
+		AbortHttpStatusRuntime: abortHttpStatusRuntime,
+		AbortPercentRuntime:    abortPercentRuntime,
+	}
+
+	var faultFilterName string
+	if isExternalFault {
+		faultFilterName = fmt.Sprintf(FaultFilterConfigNameForExternalFault, upstreamCluster)
+	} else {
+		faultFilterName = FaultFilterConfigNameForInternalFault
+
+		// match downstream cluster with request header else fault will be applied to requests coming from all downstream
+		faultFilter.Headers = []*gcpRoute.HeaderMatcher{
+			{
+				Name: EnvoyDownstreamServiceClusterHeader,
+				HeaderMatchSpecifier: &gcpRoute.HeaderMatcher_ExactMatch{
+					ExactMatch: downstreamCluster,
+				},
+			},
+		}
+	}
+
+	logger.Debugw("ECDS Fault filter", "faultFilter", faultFilter, "upstreamcluster", upstreamCluster)
+
+	serializedFaultFilter, err := proto.Marshal(faultFilter)
+	if err != nil {
+		logger.Warnw("Unable to unmarshal fault filter", "faultFilter", faultFilter)
+		return []gcpTypes.ResourceWithTtl{}
+	}
+
+	return []gcpTypes.ResourceWithTtl{{
+		Resource: &gcpCoreV3.TypedExtensionConfig{
+			Name: faultFilterName,
+			TypedConfig: &any.Any{
+				TypeUrl: faultFilterTypeURL,
+				Value:   serializedFaultFilter,
+			},
+		},
+		Ttl: ttl,
+	}}
+}
+
+func generateEmptyECDSResource(cluster string, ecdsConfig *ECDSConfig, logger *zap.SugaredLogger) []gcpTypes.ResourceWithTtl {
+	faultFilter := &gcpFilterFault.HTTPFault{
+		Delay: DefaultDelayFaultConfig,
+		Abort: DefaultAbortFaultConfig,
+
+		// override runtimes so that default runtime is not used.
+		DelayPercentRuntime:    delayPercentRuntime,
+		DelayDurationRuntime:   delayDurationRuntime,
+		AbortHttpStatusRuntime: abortHttpStatusRuntime,
+		AbortPercentRuntime:    abortPercentRuntime,
+	}
+
+	serializedFaultFilter, err := proto.Marshal(faultFilter)
```

Comment:
Should we just store this in a var somewhere? No need to do this every time if the config is identical

Reply:
By storing, do you mean a global list of `faultFilters`? I think it might be an overkill